src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOps.java

```diff
@@ -3145,46 +3142,66 @@ public static PatternOps.RecordPatternOp recordPattern(RecordTypeDesc recordDesc
     }
 
     @OpDeclaration(StringTemplateOp.NAME)
-    public static final class StringTemplateOp extends OpWithDefinition {
+    public static final class StringTemplateOp extends OpWithDefinition implements Op.Nested {
 
         public static final String NAME = "java.stringTemplate";
 
+        private final Value processorValue;
+        private final List<Value> literalsValues;
+        private final List<Body> expressionsBodies;
+
         public StringTemplateOp(OpDefinition def) {
             super(def);
+
+            this.processorValue = operands().get(0);
+            this.literalsValues = operands().subList(1, operands().size());
+            this.expressionsBodies = def.bodyDefinitions().stream().map(bd -> bd.build(this)).toList();
         }
 
         StringTemplateOp(StringTemplateOp that, CopyContext cc, OpTransformer ot) {
             super(that, cc);
+
+            this.processorValue = operands().get(0);
+            this.literalsValues = operands().subList(1, operands().size());
+            this.expressionsBodies = that.expressionsBodies.stream()
+                    .map(b -> b.transform(cc, ot).build(this)).toList();
+        }
+
+        public StringTemplateOp(Value processorValue, List<Value> literalsValues, List<Body.Builder> expressionsBodies) {
+            super(NAME, makeOperandsList(processorValue, literalsValues));
+
+            this.processorValue = processorValue;
+            this.literalsValues = literalsValues;
+            this.expressionsBodies = expressionsBodies.stream().map(b -> b.build(this)).toList();
         }
 
         @Override
         public StringTemplateOp transform(CopyContext cc, OpTransformer ot) {
             return new StringTemplateOp(this, cc, ot);
         }
 
-        public StringTemplateOp(List<Value> literalsValues, List<Value> expressionValues) {
-            super(NAME, getOperandsList(literalsValues, expressionValues));
-        }
-
-        private static List<Value> getOperandsList(List<Value> literalsValues, List<Value> expressionValues) {
+        private static List<Value> makeOperandsList(Value processorValue, List<Value> literalsValues) {
             List<Value> operands = new ArrayList<>();
-            for (int i = 0; i < literalsValues.size(); i++) {
-                operands.add(literalsValues.get(i));
-                if (i < expressionValues.size()) {
-                    operands.add(expressionValues.get(i));
-                }
-            }
+            operands.add(processorValue);
+            operands.addAll(literalsValues);
             return operands;
         }
 
         @Override
         public TypeElement resultType() {
-            return JavaType.J_L_STRING_TEMPLATE;
+            // processor type: StringTemplate$Processor<R, ...>
+            TypeDefinition processorReturnType = processorValue.type().toTypeDefinition().arguments().get(0);
+            return CoreTypeFactory.JAVA_TYPE_FACTORY.constructType(processorReturnType);
+        }
```

Comment:
That's a clever way to do it, but for now i recommend passing in the result type to the factory and constructor.

The other, and more preferred, way to operate on types is to check if the processor value's type is an instance of `JavaType` and then access as that type to obtain the argument type. We will probably revisit this whole area after we have decided how to model Java types.

Reply:
If we pass resultType, result type will be fixed even if we do transformation that change the processor, e.g. RAW -> STR

Reply:
Yeah, that's a downside. But, i think we need to revisit that issue more generally.